Support import certificate using softhsm2-util

alonbl opened the pull request:
The softhsm2-util already support importing keys, why not also import
certificates?

Useful for test scripts that require both keys and certificates.

Add --import-type <type> parameter, depreciate the --aes parameter.

Signed-off-by: Alon Bar-Lev <alon.barlev@gmail.com>

rijswijk commented:
Thank you for your contribution, I have some things to consider before merging:

Certificate import may be slightly more complex, since certificates are often associated with keys. Different applications link 
keys and certificates in different ways (e.g. one example might be to have matching `CKA_ID` attributes, but other options also exist). I'm not entirely sure whether adding this type of support to `softhsm2-util` is the right way forward, it might be too narrowly scoped to a particular use case.

@halderen your opinion would be welcome.

If we do decide to proceed with this, I think the following changes/features would be necessary:

- Don't do an immediate deprecation of ``--import``, instead, clearly mark it as deprecated in the documentation (usage, man page) and tag it for future removal (while you still support it, it is immediately deprecated without documentation)
- Support Botan as well (to keep the two builds in sync)

alonbl commented:
Hi,

Thank you for the review. I altered the implementation and added `--import-type` instead of deprecating the `--import`, I hope you like it better.

For the matching between key and certificate, as far as I know the best practice is to match the `CKA_ID` or read the entire token and match based on public key. The option to even having a basic import is significant for the use case of using the provider as test framework. If an application has special needs it can always perform the import by its own.

I have been around for years in PKCS#11 domain and never saw a setup in which the CKA_ID is not used for matching.

If you ACK I will update the man page and try to hack the botan as well.

Regards,
Alon

alonbl commented:
The botan part was actually simpler than I thought, so I applied the botan part and the man part.

rijswijk commented:
Thanks, looks good, one small request: the PR is still essentially deprecating the current mode of use of `--import` (which essentially assumes it's always a key pair unless `--aes` is specified). Furthermore, it looks like you removed the documentation for `--aes` from the manual page while the option is still silently supported. To really make it a "nice" deprecation, I suggest the following:

* If `--aes` is specified, print a warning to `stderr` that says something like "Warning: use of --aes is deprecated, this flag will be removed in a future version"
* If `--import` is specified without `--import-type`, print a warning to `stderr` that says something like "Warning: use of `--import` without specifying the object type with `--import-type` is deprecated, assuming `"keypair"`; future versions will give an error in this case."
* Clear flag the deprecated behaviour in the manual page

Again, thanks for your contribution, I think with the changes above we'd be good to go for a review by @halderen and a merge.

alonbl commented:
Hi,

I left the —aes in the man page, just removed it from the first line. And
added depreciation notice in parameter usage, so I do not think anything is
missing... just less endorsement for the parameter. :)

The past addition of the —aes was very confusing, I only realized it’s
usage when reading the code.

For the stderr messages, I do not like it, it may break existing scripts.
As long as we support backward compatibility, I thing the risk of breakage
due to output is wrong. I believe documentation should be sufficient.

What do you think?

Thanks,
Alon

rijswijk commented:
I think I agree, I'm going to ask @halderen to review and merge. Thanks!

alonbl commented:
ping?

alonbl commented:
Hi @halderen,
Can you please also checkout this one?
Thanks!

davydsantos commented:
+1 on the feature. This can be accomplished by using opensc/pkcs11-tool but it would be easier to rely on one single tool for both tasks. Another possible feature could be to allow importing a p12/pfx directly without the need of breaking it up in 2 different files and it would handle everything internally.

alonbl commented:
Rebased.

Hi @halderen,
Can you please also checkout this one? I will be glad to accept any feedback.
Thanks!

alonbl commented:
Hi @halderen,
Do you have time to review this?
Thanks.

alonbl commented:
ping?
I see almost no activity in project, any chance to merge it? feature was acked and change was reviewed.
thanks!

alonbl commented:
Hello @halderen, can you please look at this patch? It would be great if we can use the softhsm2-util to import certificates for unit tests without additional software.

alonbl commented:
Rebased.

@halderen: can you please review? this is super handy utility for unittest of PKCS\#11 enabled application.

jschlyter commented:
Please rebase on develop and mark as ready when ready.

alonbl commented:
> Please rebase on develop and mark as ready when ready.

Hi @jschlyter,
Done :)
Thanks!

ijsf commented on src/bin/util/softhsm2-util-botan.cpp:
```diff
@@ -207,6 +210,53 @@ int crypto_import_key_pair
+		{ CKA_ID,               objID,            objIDLen },
+		{ CKA_TOKEN,            &ckToken,         sizeof(ckToken) },
+		{ CKA_PRIVATE,          &ckFalse,         sizeof(ckTrue) },
+		{ CKA_VALUE,            &*blob.begin(),   (CK_ULONG)blob.size() },
```
It doesn't seem that this function is doing error checking on the user inputs and resulting vectors.

A possible result of that is that it will be dereferencing `begin()` of an empty vector, which is not allowed (https://cplusplus.com/reference/vector/vector/begin/) and would cause undefined behaviour.

ijsf replied:
This helps, thanks for changing it so quickly.

ijsf commented on src/bin/util/softhsm2-util-ossl.cpp:
```diff
@@ -216,6 +216,86 @@ int crypto_import_key_pair
+		return 1;
+	}
+
+	int blobSize = i2d_X509(x509, NULL);
```
Its advised to check if the return length (`blobSize`) is negative, in which case an error occurred (https://linux.die.net/man/3/i2d_x509).

Also for similar calls below.

alonbl replied:
I rework the implementation, I hope the cleanup pattern is acceptable.

ijsf replied:
Thanks for handling this so quickly. I know it is a bit pedantic but a check `< 0` would improve things a bit more. This is because the manual (https://linux.die.net/man/3/i2d_x509) seems to mention:
>a negative value if an error occurs
not just `-1.`

alonbl replied:
I overlooked this, fixed.

alonbl reviewed (commented):
Hello @ijsf ,

Thank you for reviewing.
I pushed a new revision, I hope it will address your concerns.

Regards,

ijsf commented on src/bin/util/softhsm2-util.cpp:
```diff
@@ -1205,6 +1248,98 @@ int importSecretKey(char* filePath, CK_SLOT_ID slotID, char* userPIN, char* labe
+	int forceExec
+)
+{
+	char user_pin_copy[MAX_PIN_LEN+1];
```
Initializing this array to zero `= {0};` will avoid problems later with uninitialized values.

alonbl replied:
done, although the `getPw()` will not return success without initializing it.

BTW: this is a copy of other import utilities, there are lots of issues in utilities including leaks.

ijsf replied:
Makes sense. Might be better to do a proper cleanup in the future.

alonbl commented:
Hello @jschlyter, is there anything more I can do to push this forward?

jschlyter commented:
I think we're good, but I'd appreciate if @bjosv could take a look as well.

bjosv reviewed (commented):
Looks useful, just have a few comment before a merge.

alonbl commented on src/bin/util/softhsm2-util-botan.cpp:
```diff
@@ -209,6 +212,64 @@ int crypto_import_key_pair
+		{ CKA_LABEL,            label,               	strlen(label) },
+		{ CKA_ID,               objID,               	objIDLen },
+		{ CKA_TOKEN,            &ckToken,            	sizeof(ckToken) },
+		{ CKA_PRIVATE,		&ckFalse,		sizeof(ckTrue) },
```
correct, done!

alonbl commented on src/bin/util/softhsm2-util-ossl.cpp:
```diff
@@ -218,6 +218,151 @@ int crypto_import_key_pair
+	{
+		CK_OBJECT_CLASS certClass = CKO_CERTIFICATE;
+		CK_CERTIFICATE_TYPE certType = CKC_X_509;
+		CK_BBOOL ckTrue = CK_TRUE, ckFalse = CK_FALSE, ckToken = CK_TRUE;
```
correct, done!

alonbl commented on src/bin/util/softhsm2-util-ossl.cpp:
```diff
@@ -218,6 +218,151 @@ int crypto_import_key_pair
+			{ CKA_LABEL,            label,        strlen(label) },
+			{ CKA_ID,               objID,        objIDLen },
+			{ CKA_TOKEN,            &ckToken,     sizeof(ckToken) },
+			{ CKA_PRIVATE,          &ckFalse,     sizeof(ckTrue) },
```
correct, done!

alonbl commented:
@bjosv thank you for spotting this leftover!

I hope all is good now.

bjosv reviewed (approved):
LGTM!

jschlyter commented:
One more thing @alonbl, could you add a test case for the import function in a separate PR?

alonbl commented:
Hello @jschlyter,

I do not see any test cases for the usage of the utilities.
I may miss this, if there is infrastructure for this, please refer me and I will add this specific test case.

Thanks,
Alon
